File: MyRadio2/app/src/main/java/com/example/myradio2/RadioService.java
package com.example.myradio2;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.media3.common.MediaItem;
import androidx.media3.common.AudioAttributes;
import androidx.media3.common.C;
import androidx.media3.exoplayer.ExoPlayer;
import androidx.media3.datasource.DefaultHttpDataSource;
import androidx.media3.exoplayer.source.MediaSource;
import androidx.media3.exoplayer.source.ProgressiveMediaSource;


public class RadioService extends Service {

    public static final String ACTION_PLAY = "ACTION_PLAY";
    public static final String ACTION_STOP = "ACTION_STOP";
    public static final String EXTRA_URL = "URL";

    private ExoPlayer player;

    @Override
    public void onCreate() {
        super.onCreate();

        startForeground(1, createNotification("초기화 중"));

        player = new ExoPlayer.Builder(this).build();

        AudioAttributes audioAttributes =
                new AudioAttributes.Builder()
                        .setUsage(C.USAGE_MEDIA)
                        .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                        .build();

        player.setAudioAttributes(audioAttributes, true);



    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        if (intent != null && ACTION_PLAY.equals(intent.getAction())) {
            String url = intent.getStringExtra(EXTRA_URL);
            playRadio(url);
        }

        return START_STICKY;
    }

    private void playRadio(String url) {

        if (player == null || url == null) return;

        android.util.Log.d("RADIO", "Play URL: " + url);

        DefaultHttpDataSource.Factory factory =
                new DefaultHttpDataSource.Factory()
                        .setUserAgent("Mozilla/5.0");

        MediaSource mediaSource =
                new ProgressiveMediaSource.Factory(factory)
                        .createMediaSource(MediaItem.fromUri(url));

        player.setMediaSource(mediaSource);
        player.prepare();
        player.play();
    }
    private Notification createNotification(String text) {

        String channelId = "radio_channel";

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel =
                    new NotificationChannel(
                            channelId,
                            "Radio Playback",
                            NotificationManager.IMPORTANCE_LOW
                    );

            NotificationManager manager =
                    getSystemService(NotificationManager.class);
            manager.createNotificationChannel(channel);
        }

        return new NotificationCompat.Builder(this, channelId)
                .setContentTitle("My Radio")
                .setContentText(text)
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .setOngoing(true)
                .build();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (player != null) {
            player.release();
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: WorldRadio/app/src/main/java/com/example/worldradio/playback/RadioService.kt
package com.example.worldradio.playback

import android.app.Service
import android.content.Intent
import android.os.IBinder
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.session.MediaSession
import androidx.media3.common.MediaItem
import android.app.NotificationChannel
import android.app.NotificationManager
import android.os.Build

/**
 * Foreground Service
 *
 * 앱이 백그라운드에 있어도
 * 음악이 중단되지 않도록 유지
 */
@AndroidEntryPoint
class RadioService : Service() {

    @Inject
    lateinit var playerManager: PlayerManager

    private lateinit var mediaSession: MediaSession

    override fun onCreate() {
        super.onCreate()

        /**
         * MediaSession 생성
         * 잠금화면/블루투스/자동차 연동
         */
        mediaSession = MediaSession.Builder(
            this,
            playerManager.player
        ).build()

        createNotificationChannel()
    }

    override fun onStartCommand(
        intent: Intent?,
        flags: Int,
        startId: Int
    ): Int {

        val url = intent?.getStringExtra("url") ?: return START_NOT_STICKY
        val name = intent.getStringExtra("name") ?: "Radio"

        playerManager.play(url)

        val notification =
            NotificationHelper.createNotification(
                this,
                mediaSession,
                name
            )

        startForeground(1, notification)

        return START_STICKY
    }

    override fun onDestroy() {
        playerManager.release()
        mediaSession.release()
        super.onDestroy()
    }

    override fun onBind(intent: Intent?) = null

    /**
     * Android 8+ 필수
     */
    private fun createNotificationChannel() {
        val channel = NotificationChannel(
            "radio_channel",
            "Radio Playback",
            NotificationManager.IMPORTANCE_LOW
        )
        getSystemService(NotificationManager::class.java)
            .createNotificationChannel(channel)
    }
}